File: parameters.py
import neatnik


# General Parameters:
# ------- ----------

# Number of generational cycles for which an `Experiment` will run.
neatnik.generational_cycles = 300

# Number of `Organism`s in an `Experiment`'s `Genus`.
neatnik.population_size = 100

# Number of attempts at mutating a `Genotype`.
neatnik.mutation_attempts = 10

# Number of attempts at producing an offspring `Organism`.
neatnik.spawning_attempts = 10

# Bounding value for the generation and replacement of `Link` weights.
neatnik.weight_bound = 2.0

# Perturbation power when altering `Link` weights.
neatnik.perturbation_power = 2.0

# Relevance of matching, disjoint, and excess `Link`s when comparing `Genotype`s.
neatnik.compatibility_weights = [0.5, 1., 1.]

# Fraction of rejected `Organism`s at a given evolution cycle.
neatnik.rejection_fraction = 0.3

# Number of evolution cycles beyond which a stagnated `Species` can be discarded.
neatnik.stagnation_threshold = 15

# The degree of similarity beyond which `Organism`s group in separate `Species`.
neatnik.compatibility_threshold = 3.0


# Probability Mass Functions:
# ----------- ---- ---------

# P.M.F. for enabling of a random DISABLED `Link`: [FAILURE, SUCCESS].
neatnik.enabling_link = [0.5, 0.5]

# P.M.F. for altering each ENABLED `Link`: [FAILURE, SUCCESS].
neatnik.altering_links = [0.5, 0.5]

# P.M.F. for altering a `Link`'s weight: [FAILURE, PERTURB, REPLACE].
neatnik.altering_weight = [0.2, 0.7, 0.1]

# P.M.F. for adding a `Link` of a given role: [FAILURE, FORWARD, RECURRENT, BIASING, LOOPED].
neatnik.adding_link = [0.5, 0.25, 0., 0.25, 0.]

# P.M.F. for enabling a DISABLED INPUT `Node`: [FAILURE, SUCCESS].
neatnik.enabling_node = [1., 0.]

# P.M.F. for altering each HIDDEN `Node`: [FAILURE, SUCCESS].
neatnik.altering_nodes = [1., 0.];

# P.M.F. for altering a `Node`'s activation: [FAILURE, HEAVISIDE, RELU, LOGISTIC].
neatnik.altering_activation = [1., 0., 0., 0.]

# P.M.F. for adding a `Node` by splitting a `Link` of a given role: [FAILURE, FORWARD, RECURRENT].
neatnik.adding_node = [0.98, 0.02, 0.]

# P.M.F. for assimilating each homologous `Link`: [FAILURE, SUCCESS].
neatnik.assimilating_links = [0., 1.]

# P.M.F. for assimilating each homologous `Node`: [FAILURE, SUCCESS].
neatnik.assimilating_nodes = [1., 0.]

# P.M.F. for assimilating a homologous `Link`'s weight: [FAILURE, SUCCESS].
neatnik.assimilating_weight = [0.5, 0.5]

# P.M.F. for assimilating a homologous `Node`'s activation function: [FAILURE, SUCCESS].
neatnik.assimilating_function = [1., 0.]

# P.M.F. for spawning a new `Organism` through a given process: [MUTATION, ASSIMILATION].
neatnik.spawning_organism = [0.4, 0.6]

File: experiment/parameters.cpp
#include "parameters.h"

// The metric responsible for driving the evolution of an `Experiment`'s `Genus`.
driver_metric Parameters::evolution_driver = FITNESS;

// Fitness threshold for `Organism`s' `Graph`s to be stored.
double Parameters::fitness_threshold = 0.;

// Novelty threshold for `Organism`s' behaviors to be stored.
double Parameters::novelty_threshold = 0.;

// Number of nearest neighbors to be considered when assessing an `Organism`'s novelty.
int Parameters::novelty_neighbors = 0;

// The factors by which the novelty threshold should increase and decrease.
std::vector<double> Parameters::novelty_threshold_modifiers = {0., 0.};

// Number of generational cycles for which an `Experiment` will run.
int Parameters::generational_cycles = 0;

// Number of `Organism`s in an `Experiment`'s `Genus`.
int Parameters::population_size = 0;

// Number of attempts at mutating a `Genotype`.
int Parameters::mutation_attempts = 0;

// Number of attempts at producing an offspring `Organism`.
int Parameters::spawning_attempts = 0;

// Bounding value for the generation and replacement of `Link` weights.
double Parameters::weight_bound = 0.;

// Perturbation power when altering `Link` weights.
double Parameters::perturbation_power = 0.;

// The activation a new HIDDEN `Node` is initially equipped with.
node_activation Parameters::initial_activation = IDENTITY;

// Fraction of rejected `Organism`s at a given evolution cycle.
double Parameters::rejection_fraction = 0.;

// Number of evolution cycles beyond which a stagnated `Species` can be discarded.
int Parameters::stagnation_threshold = 0;

// The degree of similarity beyond which `Organism`s group in separate `Species`.
double Parameters::compatibility_threshold = 0.;

// Relevance of matching, disjoint, and excess `Link`s when comparing `Genotype`s.
std::vector<double> Parameters::compatibility_weights = {0., 0., 0.};

// P.M.F. for enabling of a random DISABLED `Link`: {FAILURE, SUCCESS}.
std::vector<double> Parameters::enabling_link = {0., 0.};

// P.M.F. for altering each ENABLED `Link`: {FAILURE, SUCCESS}.
std::vector<double> Parameters::altering_links = {0., 0.};

// P.M.F. for altering a `Link`'s weight: {FAILURE, PERTURB, REPLACE}.
std::vector<double> Parameters::altering_weight = {0., 0., 0.};

// P.M.F. for adding a `Link` of a given role: {FAILURE, FORWARD, RECURRENT, BIASING, LOOPED}.
std::vector<double> Parameters::adding_link = {0., 0., 0., 0., 0.};

// P.M.F. for enabling a DISABLED INPUT `Node` and connecting it to another `Node` of a given role: {FAILURE, HIDDEN, OUTPUT}.
std::vector<double> Parameters::enabling_node = {0., 0., 0.};

// P.M.F. for altering each HIDDEN `Node`: {FAILURE, SUCCESS}.
std::vector<double> Parameters::altering_nodes = {0., 0.};

// P.M.F. for altering a `Node`'s activation: {FAILURE, HEAVISIDE, RELU, LOGISTIC}.
std::vector<double> Parameters::altering_activation = {0., 0., 0., 0.};

// P.M.F. for adding a `Node` by splitting a `Link` of a given role: {FAILURE, FORWARD, RECURRENT}.
std::vector<double> Parameters::adding_node = {0., 0., 0.};

// P.M.F. for assimilating each homologous `Link`: {FAILURE, SUCCESS}.
std::vector<double> Parameters::assimilating_links = {0., 0.};

// P.M.F. for assimilating each homologous `Node`: {FAILURE, SUCCESS}.
std::vector<double> Parameters::assimilating_nodes = {0., 0.};

// P.M.F. for assimilating a homologous `Link`'s weight: {FAILURE, SUCCESS}.
std::vector<double> Parameters::assimilating_weight = {0., 0.};

// P.M.F. for assimilating a homologous `Node`'s activation function: {FAILURE, SUCCESS}.
std::vector<double> Parameters::assimilating_activation = {0., 0.};

// P.M.F. for spawning a new `Organism` through a given process: {MUTATION, ASSIMILATION}.
std::vector<double> Parameters::spawning_organism = {0., 0.};


// Constructor:

// Default constructor responsible for initializing this `Parameters` instance.
Parameters::Parameters()
{
    // Constructs this `Parameters` instance.
}


// Destructor:

// Default destructor responsible for deleting this `Parameters` instance.
Parameters::~Parameters()
{
    // Deletes this `Parameters` instance.
}
